### PR TITLE
fix: potential overflow of coinbase calc

### PR DESCRIPTION
**Pull request by SWvheerden**
Description
---
fix potential overflow of multiple coinbase grpc call

Motivation and Context
---

How Has This Been Tested?
---
tests

### CODE REVIEW

File: applications/minotari_node/src/grpc/base_node_grpc_server.rs [hunk 1]
```diff
@@ -788,27 +788,33 @@ impl tari_rpc::base_node_server::BaseNode for BaseNodeGrpcServer {
...
+        let mut prev_coinbase_value = 0u128;
         for coinbase in &mut coinbases {
-            coinbase.value *= reward / total_shares;
-            if remainder > 0 {
-                coinbase.value += 1;
-                remainder -= 1;
-            }
+            cur_share_sum += u128::from(coinbase.value);
+            coinbase.value =
+                u64::try_from((cur_share_sum * reward) / total_shares - prev_coinbase_value).map_err(|_| {
```
Comment (AaronFeickert) [inline applications/minotari_node/src/grpc/base_node_grpc_server.rs]: Any issue with possible overflow in the multiplication, division by zero, or underflow in the subtraction?

File: integration_tests/tests/steps/node_steps.rs [hunk 2]
```diff
@@ -851,18 +851,19 @@ async fn generate_block_with_2_as_single_request_coinbases(world: &mut TariWorld
...
-            num_6154266700 += 1;
+        if output.minimum_value_promise.as_u64() == 6154272109 {
+            num_6154272109 += 1;
         }
-        if output.minimum_value_promise.as_u64() == 12308533398 {
-            num_12308533398 += 1;
+        if output.minimum_value_promise.as_u64() == 12308544218 {
+            num_12308544218 += 1;
         }
     }
```
Comment (hansieodendaal) [inline integration_tests/tests/steps/node_steps.rs:863]: These values should equal the total block reward, should it not? We could assert that.

**Review by hansieodendaal — COMMENTED** [review]
Looks good, one comment below.

File: applications/minotari_node/src/grpc/base_node_grpc_server.rs (hunk 1 above)
Comment (hansieodendaal) [reply applications/minotari_node/src/grpc/base_node_grpc_server.rs]: We could test for that, and subdue future comments.

File: integration_tests/tests/steps/node_steps.rs (hunk 2 above)
Comment (SWvheerden) [reply integration_tests/tests/steps/node_steps.rs:863]: It is checked below by the fact that the block is accepted as valid

File: applications/minotari_node/src/grpc/base_node_grpc_server.rs (hunk 1 above)
Comment (SWvheerden) [reply applications/minotari_node/src/grpc/base_node_grpc_server.rs]: over flow should not be possible as its all u64, but I made them checked and saturating anyway

**Pull request merged**